### PR TITLE
Compress merkle proofs algorithm

### PR DESCRIPTION
**Pull request by tshchelovek**
Added friend functions to merkle_proof_impl struct for generating and validating multiple proofs with shorter paths on the same tree

### CODE REVIEW

File: test/merkle/merkle.cpp [hunk 1]
```diff
@@ -93,6 +94,99 @@ void testing_validate_template(std::vector<Element> data) {
...
 }
 
+template<typename Hash, size_t Arity, typename ValueType, std::size_t N>
+void testing_validate_template_random_data_compressed_proofs(std::size_t leaf_number) {
+    std::array<ValueType, N> data_not_in_tree = {0};
+    auto data = generate_random_data<ValueType, N>(leaf_number);
+    auto tree = make_merkle_tree<Hash, Arity>(data.begin(), data.end());
+
+    std::size_t num_idxs = 3;
+    while (num_idxs == 0) {
```
Comment (SK0M0R0H) [inline test/merkle/merkle.cpp:106]: The loop content will never be called: num_idxs != 0

File: test/merkle/merkle.cpp [hunk 2]
```diff
@@ -93,6 +94,99 @@ void testing_validate_template(std::vector<Element> data) {
...
+    for (auto idx : sorted_idxs) {
+        if (idx == leaf_idx) {
+            leaf_idx++;
+        } else {
+            break;
+        }
+    }
+    auto data_wrong_leaf = data_for_validation;
+    data_wrong_leaf[std::rand() % num_idxs] = data[leaf_idx];
+    assert(data_wrong_leaf != data_for_validation);
```
Comment (SK0M0R0H) [inline test/merkle/merkle.cpp:154]: Looks like it may fail for some random cases

File: include/nil/crypto3/container/merkle/proof.hpp [hunk 3]
```diff
@@ -171,14 +172,135 @@ namespace nil {
...
+                generate_compressed_proofs(const containers::merkle_tree<typename detail::merkle_tree_node<T>::hash_type, Arity> &tree, 
+                                            std::vector<std::size_t> leaf_idxs) {
+                assert(leaf_idxs.size() > 0);
+                std::vector<std::size_t> sorted_idx(leaf_idxs.size());
+                std::iota(sorted_idx.begin(), sorted_idx.end(), 0);
+                std::sort(sorted_idx.begin(), sorted_idx.end(), [&leaf_idxs](std::size_t i, std::size_t j) {
+                                                                return leaf_idxs[i] < leaf_idxs[j]; });
+                typedef typename detail::merkle_tree_node<T> NodeType;
+                typedef typename detail::merkle_proof_impl<NodeType, Arity>::path_element_type path_element_type;
+                typedef std::array<path_element_type, Arity - 1> layer_type;
```
Comment (SK0M0R0H) [inline include/nil/crypto3/container/merkle/proof.hpp]: Why to redefine `layer_type` instead of aliasing the existing type from `merkle_proof_impl`?

File: include/nil/crypto3/container/merkle/proof.hpp [hunk 4]
```diff
@@ -171,14 +172,135 @@ namespace nil {
...
+                std::vector<detail::merkle_proof_impl<NodeType, Arity>> result_proofs(leaf_idxs.size());
+                std::size_t row_len = tree.leaves();
+                std::vector<bool> known(2 * row_len, false);
+                for (auto idx : sorted_idx) {
+                    auto leaf_idx = leaf_idxs[idx];
+                    std::vector<layer_type> path(tree.row_count() - 1);
```
Comment (SK0M0R0H) [inline include/nil/crypto3/container/merkle/proof.hpp]: `merkle_proof_impl` has path_type

File: include/nil/crypto3/container/merkle/proof.hpp [hunk 5]
```diff
@@ -171,14 +172,135 @@ namespace nil {
...
+                    typename std::vector<layer_type>::iterator v_itr = path.begin();
```
Comment (SK0M0R0H) [inline include/nil/crypto3/container/merkle/proof.hpp]: Could we rename `v_itr` to something more meaningful?

File: include/nil/crypto3/container/merkle/proof.hpp [hunk 6]
```diff
@@ -171,14 +172,135 @@ namespace nil {
...
+                    std::size_t cur_leaf = leaf_idx;
+                    std::size_t row_len = tree.leaves();
+                    std::size_t row_begin_idx = 0;
+                    bool finish_path = false;
+                    while (cur_leaf != tree.size() - 1) {
+                        std::size_t cur_leaf_pos = cur_leaf % Arity;
+                        std::size_t cur_leaf_arity_pos = (cur_leaf - row_begin_idx) / Arity;
+                        std::size_t begin_this_arity = cur_leaf - cur_leaf_pos;
+                        typename layer_type::iterator a_itr = v_itr->begin();
```
Comment (SK0M0R0H) [inline include/nil/crypto3/container/merkle/proof.hpp]: Could we rename `a_itr` to something more meaningful?

File: test/merkle/merkle.cpp [hunk 7]
```diff
@@ -153,6 +247,28 @@ BOOST_AUTO_TEST_CASE(merkletree_validate_test_3) {
...
+    testing_validate_template_compressed_proofs<hashes::sha2<256>, 3>(v);
+    testing_validate_template_compressed_proofs<hashes::md5, 3>(v);
+    testing_validate_template_compressed_proofs<hashes::blake2b<224>, 3>(v);
+
+    std::size_t leaf_number = 9;
+    testing_validate_template_random_data_compressed_proofs<hashes::sha2<256>, 3, std::uint8_t, 1>(leaf_number);
+    testing_validate_template_random_data_compressed_proofs<hashes::md5, 3, std::uint8_t, 1>(leaf_number);
+    testing_validate_template_random_data_compressed_proofs<hashes::blake2b<224>, 3, std::uint8_t, 1>(leaf_number);
+}
+
```
Comment (SK0M0R0H) [inline test/merkle/merkle.cpp:313]: No edge cases for:

- One index in the batch
- Same indexes in the batch
- Two indexes {0, leafs.size() - 1}

**Review by SK0M0R0H — CHANGES_REQUESTED** [review]

**Comment by SK0M0R0H** [comment]
Also, I'd suggest to move this functionality into the structure implementation instead of using friend methods.

File: test/merkle/merkle.cpp (hunk 7 above)
Comment (tshchelovek) [reply test/merkle/merkle.cpp:313]: Added edge cases

File: include/nil/crypto3/container/merkle/proof.hpp (hunk 6 above)
Comment (tshchelovek) [reply include/nil/crypto3/container/merkle/proof.hpp]: Renamed to layer_itr

File: include/nil/crypto3/container/merkle/proof.hpp (hunk 5 above)
Comment (tshchelovek) [reply include/nil/crypto3/container/merkle/proof.hpp]: Renamed to path_itr

File: include/nil/crypto3/container/merkle/proof.hpp (hunk 4 above)
Comment (tshchelovek) [reply include/nil/crypto3/container/merkle/proof.hpp]: Friend functions to static methods change resolved this

File: include/nil/crypto3/container/merkle/proof.hpp (hunk 3 above)
Comment (tshchelovek) [reply include/nil/crypto3/container/merkle/proof.hpp]: Friend functions to static methods change resolved this

File: test/merkle/merkle.cpp (hunk 2 above)
Comment (tshchelovek) [reply test/merkle/merkle.cpp:154]: Updated

**Comment by tshchelovek** [comment]
> Also, I'd suggest to move this functionality into the structure implementation instead of using friend methods.

Updated this and the calls of the functions in zk #107 branch

**Review by SK0M0R0H — APPROVED** [review]
